pkg/controller/keycloak/keycloak_reconciler.go

```diff
@@ -325,3 +325,36 @@ func (i *KeycloakReconciler) getPostgresqlServiceEndpointsDesiredState(clusterSt
 		Msg: "Update External Database Service Endpoints",
 	}
 }
+
+func (i *KeycloakReconciler) getPodDisruptionBudgetDesiredState(clusterState *common.ClusterState, cr *kc.Keycloak) common.ClusterAction {
+	if shouldHavePDB(cr) && clusterState.PodDisruptionBudget == nil {
+		return common.GenericCreateAction{
+			Ref: model.PodDisruptionBudget(cr),
+			Msg: "Create PodDisruptionBudget",
+		}
+	}
+
+	if shouldHavePDB(cr) && clusterState.PodDisruptionBudget != nil {
```

Comment:
`shouldHavePDB(cr)` seems unnecessary here. We perform the same action if it's `true` and `false`. I believe it can be removed altogether (the `shouldHavePDB(cr)` method).

Reply:
@slaskawi Hmm, I think `shouldHavePDB` is still needed otherwise I wouldn't be able to know what value the PDB object should be updated to. Depending on the value returned by `shouldHavePDB`, the `Ref` value is different on line L339 and L346.

Reply:
aaah right... I didn't notice that. I'm not sure if we need any logic using `MaxAvailable` or `MinAvailable` at all...

What if we used always `PodDisruptionBudgetReconciled` - the one with `MaxUnavailable`? If my understanding is correct and I didn't miss anything, this one should always work (even if `cr.Spec.Instances == 1`). If my statement is true, I think we could get rid of both `shouldHavePDB` (and rely solely on `cr.Spec.PodDisruptionBudget.Enabled`). 

One thing might also deserve some testing - what will happen, if `cr.Spec.PodDisruptionBudget.Enabled`, we used only PDB's `MaxUnavailable` and `cr.Spec.Instances == 0`. This is a valid case we use during the migration (there's a short period of time, where we ask StatefulSet to kill all the nodes). 

What do you think?

Reply:
yes, you are right, now I think about it, we can always use `PodDisruptionBudgetReconciled`, it will work regardless what number of the keycloak instances is set to. I will make the change as you suggested.

> One thing might also deserve some testing - what will happen, if cr.Spec.PodDisruptionBudget.Enabled, we used only PDB's MaxUnavailable and cr.Spec.Instances == 0. This is a valid case we use during the migration (there's a short period of time, where we ask StatefulSet to kill all the nodes).

In this [doc](https://kubernetes.io/docs/concepts/workloads/pods/disruptions/), it specifically called out about PDB and rolling upgrades:

**_"Pods which are deleted or unavailable due to a rolling upgrade to an application do count against the disruption budget, but controllers (like deployment and stateful-set) are not limited by PDBs when doing rolling upgrades – the handling of failures during application updates is configured in the controller spec. (Learn about updating a deployment.)"_**

We have tested this as well and the presents of the PDB won't stop the pods from being scaled down during upgrades.